### PR TITLE
Extend MQTT to set state / max_current / charge_current

### LINKED ISSUE #358: MQTT 'API'
We are deprecating RAPI access, at least the `$Sx` and `$Fx` commands, and replacing with a higher level HTTP API because a lot of the 'smart' feature will now be implemented on the more powerful ESP32. 

Currently none of this is exposed to MQTT, and to a large extent probably won't be as MQTT is not really designed for APIs. However there are a few use cases where controlling the EVSE via MQTT may have advantages over HTTP so the following should be made available:

- Max current
- Charge current
- Manual Override - Start/Stop charging
- Divert mode (existing API)

### PR DESCRIPTION
**Pull request by KipK**
Fixes #358

mqtt: set override ( start/stop/delete)
mqtt: publish max_current
mqtt  set charge_current + max_current

 Not sure of my choices, let me know if some approach are bad.

perhaps should I rename /charge_current/set as /pilot/set to stay coherent ?

### CODE REVIEW

**Comment by jeremypoulter** [comment]
Thanks for your contribution and tackling one of the bigger issues. I will review it shortly.

**Comment by KipK** [comment]
I have rebased it in one commit. There's probably some choices to discuss.

File: src/mqtt.cpp [hunk 1]
```diff
@@ -107,6 +108,43 @@ void mqttmsg_callback(MongooseString topic, MongooseString payload) {
...
   }
+  else if (topic_string == mqtt_topic + "/charge_current/set")
+  {
+    int newchargecurrent = payload_str.toInt();
+    DBUGF("Set charge_current: %d", newmode);
+    //using override, the only interest of this is having priority above everything else.
+    //We have a confusion in web interface to solve as details of the override are not displayed
+    EvseProperties props(EvseState::Active);
+    props.setChargeCurrent(newchargecurrent);
+    evse.claim(EvseClient_OpenEVSE_Manual, EvseManager_Priority_Manual, props);
```
Comment (jeremypoulter) [inline src/mqtt.cpp:119]: I think the priority should be the same as the HTTP API and there should probably be a specific client for these changes, eg

```suggestion
    evse.claim(EvseClient_OpenEVSE_MQTT, EvseManager_Priority_API, props);
```

File: src/mqtt.cpp [hunk 2]
```diff
@@ -107,6 +108,43 @@ void mqttmsg_callback(MongooseString topic, MongooseString payload) {
       divertmode_update(newdivert);
     }
   }
+  else if (topic_string == mqtt_topic + "/charge_current/set")
+  {
+    int newchargecurrent = payload_str.toInt();
+    DBUGF("Set charge_current: %d", newmode);
+    //using override, the only interest of this is having priority above everything else.
+    //We have a confusion in web interface to solve as details of the override are not displayed
```
Comment (jeremypoulter) [inline src/mqtt.cpp:116]: The GUI could be updated to show that the EVSE is being controlled be an external source and use the `/claims` API to get the info

File: src/mqtt.cpp [hunk 3]
```diff
@@ -107,6 +108,43 @@ void mqttmsg_callback(MongooseString topic, MongooseString payload) {
       divertmode_update(newdivert);
     }
   }
+  else if (topic_string == mqtt_topic + "/charge_current/set")
```
Comment (jeremypoulter) [inline src/mqtt.cpp:111]: This does not match the subscribed topic:

```suggestion
  else if (topic_string == mqtt_topic + "/pilot/set")
```

**Review by jeremypoulter — CHANGES_REQUESTED** [review]

File: src/mqtt.cpp (hunk 3 above)
Comment (jeremypoulter) [reply src/mqtt.cpp:111]: Also the GitHub workflows look to have been broken by Platform IO so there is a dependency on #207

**Comment by jeremypoulter** [comment]
Would also be great to add a doc for the API as `docs/mqtt.md`

File: src/mqtt.cpp (hunk 2 above)
Comment (KipK) [reply src/mqtt.cpp:116]: Agreed, that is the way. I have not gave a look yet the the javascript gui part as it's embedded in .h file, looks unreadable :)

File: src/mqtt.cpp (hunk 3 above)
Comment (KipK) [reply src/mqtt.cpp]: you mean I can merge #207 in my base main ?

**Comment by KipK** [comment]
> 

was on it ;) it's pushed

File: src/mqtt.cpp (hunk 1 above)
Comment (KipK) [reply src/mqtt.cpp:119]: I was using a dedicated MQTT id first, then removed it, to match what http interface did :)
It seems http://openevse/override use EvseClient_OpenEVSE_Manual, EvseManager_Priority_Manual ( I've copied it )

Comment (KipK) [reply src/mqtt.cpp]: I've pushed it with new EvseClient_OpenEVSE_MQTT, EvseManager_Priority_MQTT (500)

**Comment by KipK** [comment]
Also, MQTT data should be updated just after a set. For now we need to wait for the new loop to have feedback.
What would be the best way to do it in a centralised way, without rebuilding a shaped event request for each function ?

File: src/mqtt.cpp (hunk 1 above)
Comment (KipK) [reply src/mqtt.cpp]: I've tested, but it has collaterals.
Now, as priority has changed, it is not considered as an override, but just a claim.
So on web interface we can't see and delete it.
Also manual_override is not set to 1 then on mqtt. 

Should we expose claims over mqtt ? 
Do you have some suggestions ?

Comment (KipK) [reply src/mqtt.cpp]: I've opened an issue for this to not spam too much this page : https://github.com/OpenEVSE/ESP32_WiFi_V4.x/issues/373

**Comment by KipK** [comment]
I'm closing it to open a wip new one based on http API for discussion.

**Pull request closed without merging**